contracts/allocators/FraxSharesAllocator.sol

```diff
@@ -0,0 +1,172 @@
+// SPDX-License-Identifier: AGPL-3.0-or-later
+pragma solidity 0.7.5;
+
+import "@openzeppelin/contracts-upgradeable/proxy/Initializable.sol";
+import "@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/math/SafeMathUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/token/ERC20/SafeERC20Upgradeable.sol";
+
+import "../interfaces/ITreasury.sol";
+import "../interfaces/IAllocator.sol";
+import "../interfaces/IERC20.sol";
+import "../libraries/SafeERC20.sol";
+
+interface IveFXS is IERC20 {
+    /**
+     * @notice Deposit `_value` tokens for `msg.sender` and lock until `_unlock_time`
+     * @param _value Amount to deposit
+     * @param _unlock_time Epoch time when tokens unlock, rounded down to whole weeks
+     */
+    function create_lock(uint256 _value, uint256 _unlock_time) external;
+
+    /**
+     * @notice Deposit `_value` additional tokens for `msg.sender` without modifying the unlock time
+     * @param _value Amount of tokens to deposit and add to the lock
+     */
+    function increase_amount(uint256 _value) external;
+
+    /**
+     * @notice Extend the unlock time for `msg.sender` to `_unlock_time`
+     * @param _unlock_time New epoch time for unlocking 
+     */
+    function increase_unlock_time(uint256 _unlock_time) external;
+
+    /**
+     * @notice Get timestamp when `_addr`'s lock finishes
+     * @param _addr wallet address
+     * @return Epoch time of the lock end
+     */
+     function locked__end(address _addr) external view returns (uint256);
+}
+
+interface IveFXSYieldDistributorV4 {
+    /**
+     * @notice transfers FXS earned by locking veFXS
+     * @return the amount of FXS transferred
+     */
+    function getYield() external returns (uint256);
+
+    /**
+     * @notice returns the pending rewards for an address
+     */
+    function earned(address _address) external view returns (uint256);
+
+    /* BELOW USED ONLY IN TESTS */
+
+    /**
+     * @notice forces an update of a user's rewards
+     */
+    function checkpointOtherUser(address _address) external;
+
+    /**
+     * @notice requests FXS rewards to pulled from msg.sender
+     */
+    function notifyRewardAmount(uint256 amount) external;
+
+    /**
+     * @notice allows an address to call notifyRewardAmount
+     */
+    function toggleRewardNotifier(address notifier_addr) external;
+
+    /**
+     * @notice returns the number of seconds until a reward is fully distributed
+     */
+    function yieldDuration() external returns(uint256);
+}
+
+contract  FraxSharesAllocator is Initializable, OwnableUpgradeable {
+    using SafeERC20 for IERC20;
+    using SafeMathUpgradeable for uint256;
+
+    /* ======== STATE VARIABLES ======== */
+    /* !!!! UPGRADABLE CONTRACT !!!! */
+    /* NEW STATE VARIABLES MUST BE APPENDED TO END */
+
+    uint256 constant private MAX_TIME = 4 * 365 * 86400 + 1;  // 4 years and 1 second
+    ITreasury public treasury;
```

Comment:
So since FXS is in the v1 treasury, can we add a setter to this? should allow for EZ switching to v2 treasury when we done.

Reply:
Donezo